### ISSUE TITLE
Allow building only dependencies of a target

### ISSUE THREAD

**Issue by rlepigre** [ocaml/dune#7972]
## Desired Behavior

Colleagues of mine have often asked me if there is a way to tell dune to build only the dependencies of some target. This has come up again today in the context of a Coq project, where we have relied on the following hack.

To build all the dependencies of a Coq file (which is useful to then load it into the Coq toplevel), on can use `dune coq top dir/file.v`. This has the same effect as building all the dependencies for file `dir/file.vo`, without building `dir/file.vo` itself.

However, this approach is limited to cases where you want to build the deps for a single file. It would not be reasonable to allow several file parameters to `dune coq top`, since that does not really make any sense from the point of view for the toplevel.

A better option would be to add a `--deps-only` argument to `dune build`, so that one can run, for example, `dune build --deps-only dir/file1.vo dir1/file2.vo` to build all the dependencies of `dir/file1.vo` and `dir/file2.vo`.

**Comment by ejgallego** [comment]
Hi @rlepigre , can you tell more about why you need this (other than calling coqtop) ?

**Comment by ejgallego** [comment]
I'm unsure how hard is to implement a general `dune build --deps_only $target` command, it could likely require patching the engine.

For `dune coq top` as you well know we actually expose the `deps_of` build rule, so that is "ad-hoc" in the sense of rules.

**Comment by rlepigre** [comment]
Could we easily have an alias that builds the dependencies of a given Coq file? (Not sure how these work.) If we could do something like `dune build path/to/file.v@deps` (or whatever the syntax is supposed to be) that would definitely be good enough for us.

**Comment by aa755** [comment]
this is independent of coqtop. we use shared dune caches between local development machines and CI machines, using the same docker image. once CI builds the currrent state, we get the CI error log and would like to just build (actually, just fetch from cache) the dependencies of the files that indicate error in the CI log. then the user can edit the errored files in whatever editor they log.

**Comment by Alizter** [comment]
FTR we already have ways to build only deps exposed in the engine and all that is needed is the plumbing for the bin/ commands. There is some processing of targets that needs to be circumvented and replaced with a dependencies query which will be fed into `Build_system.build_deps`. I am still not fully sold on the use case however.

**Comment by ejgallego** [comment]
> this is independent of coqtop. we use shared dune caches between local development machines and CI machines, using the same docker image. once CI builds the currrent state, we get the CI error log and would like to just build (actually, just fetch from cache) the dependencies of the files that indicate error in the CI log. then the user can edit the errored files in whatever editor they log. the alternative is to do `dune coq top` for each errored file, which has several issues: 1) cannot run them in parallel 2) parsing the rule again for each errrored files can waste time

Indeed, these seem like real issues. I think the easiest short-term solution would be to extend `dune coq top` to take multiple `.v` files as arguments, that should solve both issues 1) and 2)

The semantics would be to pass all the `.v` files as arguments to the chosen toplevel (in this case you would use `true` as the toplevel)

> Could we easily have an alias that builds the dependencies of a given Coq file? (Not sure how these work.) If we could do something like `dune build path/to/file.v@deps` (or whatever the syntax is supposed to be) that would definitely be good enough for us.

I think so, however given @Alizter feedback I still think a more general `--deps_only` flag could work fine too, and would allow us to remove all the `deps_of` export code we had to implement for `dune coq top`. Instead `dune coq top` could just fetch the rule for the particular target file and call `build_deps`.
